AP-824, AP-821: Amplify Auth integration

ap-justin opened the pull request:
Ticket(s):
-

Merged: 
* #2431

NOTES: 
- user is not yet tied to a particular endowment.

## Explanation of the solution
* create `/login` page
* customize amplify `<Authenticator/>` component to match ap colors
<img width="1031" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/e0c91bb9-4361-47f8-ad9f-076040879f6f">

* add `login` button in header
<img width="919" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/321730a3-6c70-4c4e-92e6-0f0a0af04fd7">

* logged in user would see header like this ( would be populated later on ) 
<img width="473" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/f8096016-7c3c-48ed-a7fb-a3bf3619396f">


* make `/admin` a protected route

* add logout button in `/admin`
<img width="671" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/59b3c732-9a9a-4ca3-aee8-54d9daa16db6">



## Instructions on making this work
* go to `admin/8` manually
* will be redirected to `/login`  to sign-in or create account
* after sign-in/sign-up, redirected back to `admin/8`

- run `yarn` or `yarn install` to install npm dependencies
- run `yarn run test --watchAll` to verify all tests still pass
- (optional) run `yarn run build` to verify the build passes
- run `yarn start` to start the webapp
-

## UI changes for review

When major UI changes are introduced with a PR, please include links to URLS to compare or screenshots demonstrating the difference and notify on design changes

ap-justin commented on .gitignore:
```diff
@@ -33,3 +33,25 @@ yarn-error.log*
+build/
+dist/
+node_modules/
+aws-exports.js
```
@SovereignAndrey , since `aws-exports` is not included in our source control , CI encounters this error
`Module not found: Error: Can't resolve './aws-exports' in '/home/runner/work/angelprotocol-web-app/angelprotocol-web-app/src'`

I tried to add some commands to `amplify.yaml` per docs
but doesn't work :( 

maybe need this step ? 
https://docs.aws.amazon.com/amplify/latest/userguide/amplify-config-autogeneration.html

SovereignAndrey replied:
@ap-justin Hmm... I thought we  had the continuous deployments and builds setup. Will check if it's not an issue with IAM Role for Amplify. :+1:

SovereignAndrey replied:
@ap-justin OK! I added `AmplifyBackendDeployFullAccess` to the existing Amplify Service Role that is assigned. See if that works for ya. :crossed_fingers:

SovereignAndrey replied:
Ah, I think I understand now. It's the GH CI that's the issue, now AWS. We will prob need to setup a GH CI user. On it!

misicnenad commented:
@ap-justin 
> add logout button in `/admin`

Shouldn't users on `/admin` page still have access to the usual "sign out" button in the top right of the header? Maybe I'm missing something.
![image](https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/19427053/1ccba8c2-1553-4400-9fbf-e81373a826b2)

misicnenad commented on src/App/Header/UserMenu/Menu.tsx:
```diff
@@ -0,0 +1,22 @@
+      <p className="text-sm p-2 text-gray-d1 dark:text-gray">
+        {user?.attributes?.email}
+      </p>
+      <div className=""></div>
```
This a placeholder for future use?

misicnenad commented on src/App/Header/UserMenu/Menu.tsx:
```diff
@@ -0,0 +1,22 @@
+      <p className="text-sm p-2 text-gray-d1 dark:text-gray">
+        {user?.attributes?.email}
+      </p>
+      <div className=""></div>
```
This a placeholder for future use?

ap-justin replied:
we could remove this for now

misicnenad reviewed (approved):
Seems good

ap-justin commented:
> @ap-justin
> 
> > add logout button in `/admin`
> 
> Shouldn't users on `/admin` page still have access to the usual "sign out" button in the top right of the header? Maybe I'm missing something. ![image](https://user-images.githubusercontent.com/19427053/275396712-1ccba8c2-1553-4400-9fbf-e81373a826b2.png)

you are right ser, the logout button in admin is there when I have not included login in the header - would remove now

misicnenad commented on src/App/Header/UserMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,24 @@
+        </Popover.Button>
+      ) : (
+        <Popover.Button className="btn-orange px-3 h-10 rounded-lg text-sm">
+          Login
```
Suggestion: _we should use consistent naming_, i.e. if this is called "Login", then the opposite functionality should be called "Logout" (as opposed to the current "Sign out") or vice versa ("Sign in" + "Sign out").
Considering Amplify uses the latter terminology, it would make sense to use the same one

SovereignAndrey replied:
Agreed! That's prob my fault in writing up the specs. :bow: "Sign up" and "Sign In" is most consistent. :+1:

ap-justin replied:
@misicnenad, @SovereignAndrey - done 
![image](https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/c654a1a7-910d-42bc-bdaf-f562c94718d2)

misicnenad commented on src/App/Header/UserMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,24 @@
+import Menu from "./Menu";
+
+export default function UserMenu() {
+  const { user, signOut } = useAuthenticator((context) => [context.route]);
```
Found a bug, repro steps below:
1. Open webapp (Marketplace)
2. Click Login button
3. Create profile if non-existent
4. Once profile is created, sign in
5. Reload page
6. Verify that you're still logged in
7. Click on the profile icon in top right (to sign out)
8. Click "Sign Out" button
Expected: user logged out
Actual: nothing happens
9. Reload page again -OR- manually type in a new URL in the URL bar
10. User is now logged out for real

Repro video below (first part shows regular sign-in/sign-out, the second part shows the bug)
(ignore Loom notifications):
https://www.loom.com/share/71503d8f99e24168aa8a6ae44da89801?sid=ac5ff299-b82e-4d2c-a0e8-66c0eba5cdfb

ap-justin replied:
it seems `useAuthenticator.user` is not good indicator if user is logged in.
fixed https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2427/commits/736ebd64e871d995151f5e4c9776a862d68edaca

misicnenad commented on src/App/Header/UserMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,24 @@
+import Menu from "./Menu";
+
+export default function UserMenu() {
+  const { user, signOut } = useAuthenticator((context) => [context.route]);
```
Bug -  if a user uses the top-right "LOGIN" button for registration, the user appears to be automatically signed in, but when they immediately click on the profile icon in the top right and try to sign out, nothing happens. Furthermore, the profile icon (indicating user is logged in) is present even when navigating to other pages, unless the user navigates by typing the address in the URL bar.

Expectation: when using the top right LOGIN button for registration, the user should not appear to be signed in (when in fact they're not) and they should be immediately asked to confirm their email. This requirement needs to be indicated to the user somehow.

Repro video below (ignore Loom notifications):
https://www.loom.com/share/eb1e1f1752014ea39a16bb96103de8d1?sid=ec35b1b0-db94-4771-8c10-34d63bb5e37f

misicnenad commented on src/App/Header/UserMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,24 @@
+import Menu from "./Menu";
+
+export default function UserMenu() {
+  const { user, signOut } = useAuthenticator((context) => [context.route]);
```
(Possible) Bug - if a user uses the top-right "LOGIN" button for registration, once they register they *seem* to be signed in even w/o confirming their email, but when they navigate to the Admin Panel, they are asked to sign in again and when they do then they are asked to confirm their email.

Expectation: the user should be asked to confirm their email immediately on registration even when using the top right LOGIN button. This requirement needs to be indicated to the user somehow.

Repro video below (ignore Loom notifications):
https://www.loom.com/share/6c8100dfcb664bf58b07da298011da36?sid=e08239e7-e0c6-4989-81c1-80e75e797457

ap-justin replied:
this is also fixed by 
https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2427/commits/736ebd64e871d995151f5e4c9776a862d68edaca

SovereignAndrey commented on src/App/App.tsx:
```diff
@@ -19,6 +21,7 @@ const Donate = lazy(() => import("pages/Donate"));
 const DonateFiatThanks = lazy(() => import("pages/DonateFiatThanks"));
 const Gift = lazy(() => import("pages/Gift"));
 const DonateWidget = lazy(() => import("pages/DonateWidget"));
+const Login = lazy(() => import("pages/Login"));
```
Let's also keep the internal naming and URL slug the same too, using "Signin".
```suggestion
const Signin = lazy(() => import("pages/Signin"));
```

SovereignAndrey commented on src/App/App.tsx:
```diff
@@ -58,7 +68,7 @@ export default function App() {
               <Route path={`${appRoutes.gift}/*`} element={<Gift />} />
               <Route path={appRoutes.marketplace} element={<Marketplace />} />
-
+              <Route path={appRoutes.login} element={<Login />} />
```
```suggestion
              <Route path={appRoutes.signin} element={<Signin />} />
```

ap-justin replied:
done

SovereignAndrey commented on src/App/Header/index.tsx:
```diff
@@ -45,9 +45,9 @@ export default function Header({ classes, links, logo }: Props) {
+        <div className="flex gap-4 justify-self-end items-center">
           <ThemeToggle classes="hidden lg:flex" />
-          <WalletSuite />
+          {location.pathname !== appRoutes.login && <UserMenu />}
```
```suggestion
          {location.pathname !== appRoutes.signin && <UserMenu />}
```

SovereignAndrey commented on src/App/Protected.tsx:
```diff
@@ -0,0 +1,14 @@
+  const location = useLocation();
+  const { route } = useAuthenticator((context) => [context.route]);
+  if (route !== "authenticated") {
+    return <Navigate to={appRoutes.login} state={{ from: location }} replace />;
```
```suggestion
    return <Navigate to={appRoutes.signin} state={{ from: location }} replace />;
```

SovereignAndrey commented on src/components/Icon/icons.ts:
```diff
@@ -106,6 +109,7 @@ export const icons = {
   LinkedinIn: FaLinkedinIn,
   ListBox: IoMdListBox,
   Loading: VscLoading,
+  Logout: HiLogout,
```
```suggestion
  Signout: HiLogout,
```

ap-justin replied:
removed this icon instead as it is unused

SovereignAndrey commented on src/components/Icon/icons.ts:
```diff
@@ -97,6 +99,7 @@ export const icons = {
   FilterLeft: BsFilterLeft,
   Giftcard: MdCardGiftcard,
   Globe: BiGlobe,
+  Google: AiFillGoogleCircle,
```
This is not utilised yet that I can find in the code, though it _will_ be for the login with Google button, yes?

ap-justin replied:
yes this was the remnant of the custom UI form - would delete this

SovereignAndrey commented on src/pages/Admin/Sidebar/LogoutBtn.tsx:
```diff
@@ -0,0 +1,16 @@
+      className="flex items-center gap-2 text-left text-gray-d2 dark:text-gray hover:text-orange dark:hover:text-orange px-5 border-t border-prim font-bold text-sm py-4"
+    >
+      <Icon type="Logout" style={{ transform: "rotateY(180deg)" }} size={20} />
+      <span>Logout</span>
```
```suggestion
      <span>Sign out</span>
```

ap-justin replied:
this <LogoutBtn/> component was deleted in favor of the header logout btn

SovereignAndrey reviewed (commented):
@ap-justin Looks good for this basic setup! The follow requirements will still be needed in: 
- [ ] Add the "login with Google" button (at a minimum) 

We'll also probably need to make changes to the Congnito interface a bit with User Pool changes (ex. collect a first / last name on signup) in order to added to personalise the UI, but will try to confirm this with product team before we do that. :+1: 

PS. I added access to Cognito to your AWS user profile, should you need to poke around and play with the User Pools

misicnenad commented on src/App/Header/UserMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,31 @@
+        </Popover.Button>
+      ) : (
+        <Popover.Button className="btn-orange px-3 h-10 rounded-lg text-sm">
+          Sign in
```
"Verify email" panel colors are off in "dark mode", text should be white
![image](https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/19427053/f580f852-3609-4547-b51e-8c40df62991b)

ap-justin replied:
fixed https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2427/commits/44b41584072149fb4d20d23b7b9f565ece2abc23

misicnenad commented on src/App/Header/UserMenu/UserMenu.tsx:
```diff
@@ -0,0 +1,31 @@
+      ) : (
+        <Popover.Button className="btn-orange px-3 h-10 rounded-lg text-sm">
+          Sign in
+        </Popover.Button>
```
When a new profile is created using the top right "Sign in" button, is it possible to add some kind of loading indicator after the email verification step? Not sure is amplify allows for such a thing, e.g. if it has an `isLoading` flag.

Repro video below. The loading indicator I'm asking about would be placed after the email verification step (at 0:14). Without this indicator it would seem to the user as if nothing is happening and then a couple of seconds later the "Sign in" button changes to the profile icon and they're suddenly signed in:
https://www.loom.com/share/620b74b2c42c4f05b160a27d7958d229?sid=b186d5a6-5637-4e23-a99d-64c77c7dbce4

ap-justin replied:
fixed https://github.com/AngelProtocolFinance/angelprotocol-web-app/pull/2427/commits/b84421be5352dbd1e00fd68794123013352930eb

ap-justin commented:
> @ap-justin Looks good for this basic setup! The following requirements are still be needed:
> 
> * [ ]  Add the "login with Google" button (at a minimum)
> 
> We'll also probably need to make changes to the Congnito interface a bit with User Pool changes (ex. collect a first / last name on signup) in order to added to personalise the UI, but will try to confirm this with product team before we do that. 👍
> 
> PS. I added access to Cognito to your AWS user profile, should you need to poke around and play with the User Pools

@SovereignAndrey , added google as option, though the google oauth clientid/secret is from my dummy account. 
to configure using ap account, I've outlined here some steps
1. go to `https://console.cloud.google.com/apis/credentials`
2. create new project
3. create oAuthconsent screen
4. select basic scope
<img width="737" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/b197619b-4fca-411a-a032-a436fc2c8ce3">


5. create credentials to get `clientID/secret`
6. in amplify studio/authentication add these secrets, and set app domain as redirect URL
<img width="1055" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/97bd3f17-5276-40f6-90eb-e888a0697fc1">

7. deploy and paste this redirect URL 
<img width="1041" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/e286d7aa-d467-43b0-ac2b-8b31ce071566">
in google developer portal
<img width="1007" alt="image" src="https://github.com/AngelProtocolFinance/angelprotocol-web-app/assets/89639563/377d01ba-c0d9-4148-a76c-14a463609a6e">


8. after these setup, changes are needed to be pulled in web-app
`amplify pull --appId dpspevs7tj1ov --envName dev`

SovereignAndrey commented:
> > @ap-justin Looks good for this basic setup! The following requirements are still be needed:
> > 
> > * [ ]   Add the "login with Google" button (at a minimum)
> > 
> > We'll also probably need to make changes to the Congnito interface a bit with User Pool changes (ex. collect a first / last name on signup) in order to added to personalise the UI, but will try to confirm this with product team before we do that. 👍
> > PS. I added access to Cognito to your AWS user profile, should you need to poke around and play with the User Pools
> 
> @SovereignAndrey , added google as option, though the google oauth clientid/secret is from my dummy account. <img alt="image" width="577" src="https://user-images.githubusercontent.com/89639563/275821724-4070e053-cd03-44bf-b0d4-214d3c220904.png">
> 
> to configure using ap account, I've outlined here some steps
> 
> 1. go to `https://console.cloud.google.com/apis/credentials`
> 2. create new project
> 3. create oAuthconsent screen
> 4. select basic scope
> 
> <img alt="image" width="737" src="https://user-images.githubusercontent.com/89639563/275820406-b197619b-4fca-411a-a032-a436fc2c8ce3.png">
> 5. create credentials to get `clientID/secret`
> 6. in amplify studio/authentication add these secrets, and set app domain as redirect URL
> 
> <img alt="image" width="1055" src="https://user-images.githubusercontent.com/89639563/275821013-97bd3f17-5276-40f6-90eb-e888a0697fc1.png">
> 7. deploy and paste this redirect URL
> 
> <img alt="image" width="1041" src="https://user-images.githubusercontent.com/89639563/275821278-e286d7aa-d467-43b0-ac2b-8b31ce071566.png"> in google developer portal <img alt="image" width="1007" src="https://user-images.githubusercontent.com/89639563/275821356-377d01ba-c0d9-4148-a76c-14a463609a6e.png">
> 8. after these setup, changes are needed to be pulled in web-app
>    `amplify pull --appId dpspevs7tj1ov --envName dev`

@ap-justin I had a GCP OAuth setup configured for AP on an old user pool. Will just copy that over to the new one we have for `dev`. Stand by! :)

misicnenad reviewed (approved):
:+1:
